File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/graph/Problem1260.java
package com.keunwon.algorithm.graph;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem1260 {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

    private static int N, M, V;
    private static boolean[][] graph;
    private static boolean[] visited;

    public static void main(String[] args) throws IOException {
        input();
        dfs(V);
        reset();
        bfs(V);
    }

    private static void bfs(int num) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(num);
        visited[num] = true;

        while (!queue.isEmpty()) {
            final int value = queue.poll();
            System.out.print(value + " ");

            for (int i = 1; i <= N; i++) {
                if (graph[value][i] && !visited[i]) {
                    queue.offer(i);
                    visited[i] = true;
                }
            }
        }
    }

    private static void dfs(int num) {
        visited[num] = true;
        System.out.print(num + " ");

        for (int i = 1; i <= N; i++) {
            if (graph[num][i] && !visited[i]) {
                dfs(i);
            }
        }
    }

    private static void reset() {
        System.out.println();
        Arrays.fill(visited, false);
    }

    private static void input() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        V = Integer.parseInt(st.nextToken());
        graph = new boolean[N + 1][N + 1];
        visited = new boolean[N + 1];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            final int num1 = Integer.parseInt(st.nextToken());
            final int num2 = Integer.parseInt(st.nextToken());

            graph[num1][num2] = graph[num2][num1] = true;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem1260.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

class Problem1260 {
    private lateinit var graph: Array<MutableList<Int>>

    private val answer = Array(2) { mutableListOf<Int>() }

    fun solution(n: Int, v: Int, edges: Array<Pair<Int, Int>>): Array<IntArray> {
        this.graph = Array(n + 1) { mutableListOf<Int>() }.apply {
            edges.forEach { (a, b) ->
                this[a].add(b)
                this[b].add(a)
            }
        }
        graph.forEach { it.sort() }

        dfs(v, BooleanArray(n + 1))
        bfs(v, BooleanArray(n + 1))
        return answer.map { it.toIntArray() }.toTypedArray()
    }

    private fun dfs(cur: Int, visited: BooleanArray) {
        if (visited[cur]) return

        visited[cur] = true
        answer[0].add(cur)

        for (next in graph[cur]) {
            if (!visited[next]) dfs(next, visited)
        }
    }

    private fun bfs(v: Int, visited: BooleanArray) {
        val queue = LinkedList<Int>().apply { offer(v) }
        visited[v] = true

        while (queue.isNotEmpty()) {
            val cur = queue.poll()
            answer[1].add(cur)

            for (next in graph[cur]) {
                if (!visited[next]) {
                    visited[next] = true
                    queue.offer(next)
                }
            }
        }
    }
}

fun main() {
    val (n, m, v) = readln().split(" ").map { it.toInt() }
    val edges = Array(m) {
        val st = StringTokenizer(readln())
        Pair(st.nextToken().toInt(), st.nextToken().toInt())
    }

    Problem1260().solution(n, v, edges).forEach { println(it.joinToString(" ")) }
}
